File: report/src/main/java/com/vmlens/report/runelementtype/memoryaccesskey/FieldIdAndObjectHashcode.java
package com.vmlens.report.runelementtype.memoryaccesskey;

import com.vmlens.report.description.DescriptionContext;
import com.vmlens.report.description.NeedsDescriptionCallback;

public class FieldIdAndObjectHashcode implements MemoryAccessKey {

    private final int fieldId;
    private final long objectHashCode;

    public FieldIdAndObjectHashcode(int fieldId, long objectHashCode) {
        this.fieldId = fieldId;
        this.objectHashCode = objectHashCode;
    }

    @Override
    public String asString(DescriptionContext context) {
        return context.fieldName(fieldId) ;
    }

    @Override
    public void addToNeedsDescription(NeedsDescriptionCallback callback) {
        callback.needsField(fieldId);
    }

    @Override
    public String objectHashCode() {
        return ""  + objectHashCode;
    }
}

File: sync-bug/src/main/scala/com/anarsoft/race/detection/report/element/runelementtype/memoryaccesskey/FieldIdAndObjectHashcode.scala
package com.anarsoft.race.detection.report.element.runelementtype.memoryaccesskey

import com.anarsoft.race.detection.report.description.{DescriptionContext, NeedsDescriptionCallback}
import com.anarsoft.race.detection.report.element.runelementtype.dominatormemoryaccesskey.GenericDominatorMemoryAccessKey


class FieldIdAndObjectHashcode(val fieldId: Int,val objectHashCode: Long) 
  extends GenericDominatorMemoryAccessKey[FieldIdAndObjectHashcode] with MemoryAccessKey  {
  override def asString(context: DescriptionContext): String = context.fieldName(fieldId)

  override def addToNeedsDescription(callback: NeedsDescriptionCallback): Unit = {
    callback.needsField(fieldId)
  }

  override def objectHashCodeOption: Option[Long] = Some(objectHashCode);


  private def canEqual(other: Any): Boolean = other.isInstanceOf[FieldIdAndObjectHashcode]
  
  override def equals(other: Any): Boolean = other match {
    case that: FieldIdAndObjectHashcode =>
      that.canEqual(this) &&
        fieldId == that.fieldId &&
        objectHashCode == that.objectHashCode
    case _ => false
  }
  
  override def hashCode(): Int = {
    val state = Seq(fieldId, objectHashCode)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }

  override def compare(that: FieldIdAndObjectHashcode): Int = {
    if (objectHashCode != that.objectHashCode) {
      objectHashCode.compare(that.objectHashCode)
    } else {
      fieldId.compare(that.fieldId)
    }
  }
}
